set/name_trie.go

```diff
@@ -0,0 +1,96 @@
+package set
+
+import (
+    "sort"
+)
+
+type NameTrie struct {
+    children   map[rune]*NameTrie
+    terminates bool
+}
+
+func createNameTrie() *NameTrie {
+    return &NameTrie{make(map[rune]*NameTrie), false}
+}
+
+func (tree *NameTrie) Exists(name string) bool {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    return ok && node.terminates
+}
+
+func (tree *NameTrie) Insert(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, true)
+    if ok {
+        node.terminates = true
+    }
+}
+
+func (tree *NameTrie) Delete(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    if ok {
+        node.terminates = false
+    }
+}
+
+func (node *NameTrie) OrderedChildren() []rune {
+    keys := []string{}
+    for suffix := range(node.children){ keys = append(keys, string(suffix)) }
+    sort.Strings(keys)
+    sorted_runes := []rune{}
+    for i := range(keys){ sorted_runes = append(sorted_runes, rune(keys[i][0]))}
+    return sorted_runes
+}
+
+func (tree *NameTrie) traverse(remainder []rune, create bool) (nextTree *NameTrie, ok bool) {
+    nextRune := remainder[0]
+    nextTree, ok = tree.children[nextRune]
+    if !ok {
+        if create {
+            tree.children[nextRune] = createNameTrie()
+            nextTree = tree.children[nextRune]
+        } else {
+            return createNameTrie(), false
+        }
+    }
+    if len(remainder) < 2 {
+        return nextTree, true
+    } else {
```

Comment:
Style: Could get rid of the else and just have the return on its own after. This helps signal what the "default flow" is when reading the code.